platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/serialize/DefaultGlobalValueCodec.kt

```diff
@@ -107,7 +109,7 @@ class DefaultSharedObjectDecoder(
 
         fun get(): Any {
             val state = state.get()
-            if (value == null && state < ReaderState.STOPPED && !latch.await(1, TimeUnit.MINUTES)) {
```

Comment:
> The check for `value == null` was to improve the failure mode in case the value was not loaded but the reader already stopped (due to some bug in the implementation).

I don't see how, though. Either way you'll fail or pass the `require` below. Actually, short-circuiting for `Reader.STOPPED` also causes the racy read and should be fixed.

Reply:
You are right, I think it was from before I had the check for `state < ReadyState.STOPPED`.  

I don't follow this though:

> short-circuiting for Reader.STOPPED also causes the racy read and should be fixed

Which race condition are you thinking of?

Also, shouldn't `value` be made thread-safe, maybe by turning it into an `AtomicReference`?

Reply:
>Which race condition are you thinking of?

We have one thread writing `value` and doing `countDown`, and another one doing `await` and reading `value`. This is safe because there is a ["happens-before"](https://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.4.5) relation between `countDown` and `await`, so it is guaranteed that the `value` is safe for reading. However, if we don't call `await`, this "happen-before" isn't there anymore, and reading the `value` may be unsafe (but looks like it isn't).

>shouldn't value be made thread-safe, maybe by turning it into an AtomicReference?

Not necessarily, because we have other synchronization primitives to introduce a happens-before edge. I looked deeper into the reader state, and all value writes happen-before writing `STOPPED`, which happens-before reading it here. So reading the value should be safe now. Though I'm not a big fan of deducing a happens-before chain this long, maybe making the `value` volatile would be more robust.